src/promptflow-azure/promptflow/azure/_storage/cosmosdb/span.py

```diff
@@ -42,21 +46,43 @@ def __init__(self, span: SpanEntity, collection_id: str, created_by: dict) -> No
         self.collection_id = collection_id
         self.id = span.span_id
         self.created_by = created_by
+        self.external_event_data_uris = []
 
-    def persist(self, client):
+    def persist(self, cosmos_client: ContainerProxy, blob_container_client: ContainerClient, blob_base_uri: str):
         if self.id is None or self.partition_key is None or self.resource is None:
             return
 
+        if self.events and blob_container_client is not None and blob_base_uri is not None:
+            self._persist_events(blob_container_client, blob_base_uri)
+
         resource_attributes = self.resource.get(SpanFieldName.ATTRIBUTES, None)
         if resource_attributes is None:
             return
 
         from azure.cosmos.exceptions import CosmosResourceExistsError
 
         try:
-            return client.create_item(body=self.to_dict())
+            return cosmos_client.create_item(body=self.to_dict())
         except CosmosResourceExistsError:
             return None
 
     def to_dict(self) -> Dict[str, Any]:
         return {k: v for k, v in self.__dict__.items() if v}
+
+    def _persist_events(self, blob_container_client: ContainerClient, blob_base_uri: str):
+        if not blob_base_uri.endswith("/"):
+            blob_base_uri += "/"
+
+        for idx, event in enumerate(self.events):
+            event_data = json.dumps(event)
+            blob_client = blob_container_client.get_blob_client(self._event_path(idx))
+            blob_client.upload_blob(event_data)
```

Comment:
Shall we handle error for upload_blob?

Reply:
I think we should throw exception here? Because we don't know how to handle blob storage fail currently.

Reply:
OK